src/widgets/block.rs

```diff
@@ -1134,64 +1120,19 @@ mod tests {
     const fn block_can_be_const() {
         const _DEFAULT_STYLE: Style = Style::new();
         const _DEFAULT_PADDING: Padding = Padding::uniform(1);
-        const _DEFAULT_BLOCK: Block = Block::new()
+        const _DEFAULT_BLOCK: Block = Block::bordered()
             // the following methods are no longer const because they use Into<Style>
             // .style(_DEFAULT_STYLE)           // no longer const
             // .border_style(_DEFAULT_STYLE)    // no longer const
             // .title_style(_DEFAULT_STYLE)     // no longer const
             .title_alignment(Alignment::Left)
             .title_position(Position::Top)
-            .borders(Borders::ALL)
             .padding(_DEFAULT_PADDING);
     }
 
-    /// This test ensures that we have some coverage on the [`Style::from()`] implementations
-    #[test]
-    fn block_style() {
```

Comment:
This test is an intentional duplication of the tests in style, acting as capturing the intent of the style implementation. It exists because it's rare we'd call to `Style::from()` manually, and much more common to see the `.into()` call that this exercises.

While it's true that in effect this tests that the blanket implementation of `Into` is implemented on all the items that have a `From`, that's the effect that we want.

Put more simply, we want to ensure that a user can configure various widgets using styles that are automatically converted like this. That, not the `Style::from()` call, is the important thing to test.

This could probably do with a bit more of an expanded doc comment that helps show the rationale for this. Perhaps:

/// This test intentionally duplicates the tests of `Style::from()` implementations. It does so to test from the perspective of the common use case (passing various values to a method that accepts `Into<Style>`. It captures the intent and documents the valid values better than the corresponding Style unit tests.

Reply:
Then its still a test for Style and should be with the style implementation and style tests?

Reply:
When someone needs to read tests to understand / find / use this, then its probably bad documentation or bad design.

`Into` creates hard to understand code. What exactly is happening? (To me personally it also created performance issues in the past without being obvious as the actual implementation is hidden a few steps away and did things I wasn't aware of.)

I can assume that your example is foreground, background, modifier, but it's not clearly stated resulting in guessing when reading source code. Yes it might be slightly shorter, but we have modern editors with autocompletion. `Style::new().fg().bg().modifier()` or `Block::new().fg().bg().modifier()` isn't that much longer but so much easier to read. So personally I try to avoid Into in the first place. (Into is one of the biggest things I don't like about the ratatui code design. But as long as I don't need to use it extensively…)

Tests are there to make sure it works. Examples are there to display how something could be done. Docs to highlight specific usage of some type. Maybe move this to the examples or docs then?

Reply:
I'd really like to keep these tests, but I don't mind moving them into Style instead of block or into examples / docs - it's hard to know what the right place for this is (hence defaulting to just putting it in the block tests).

The biggest reason behind keeping this is that it tests the usage of the shortcuts from the perspective of a user - i.e. what the applications that people write will see. If these tests break, then we know that there are customer apps that will be broken. Testing From says the same thing, but when those break we know that customer code that is different from the tested code breaks. I prefer a direct first order effect over a second order one when possible.

Regarding the use of `Into`, `some_value<T: Into<SomeValue>>(some_value: T)` approach is something that a bunch of the derive builder crates tend to do, and this is just a manual version of that.
- https://docs.rs/derive_builder/latest/derive_builder/#generic-setters
- https://idanarye.github.io/rust-typed-builder/typed_builder/derive.TypedBuilder.html#customization-with-attributes
- https://crates.io/crates/derive_setters
- etc.

I think there's probably a few places where we probably should have used conversion implementations rather than Into in a builder parameter (specifically in constructors). There's also some places where we should fix up the From Vec stuff to make this supper FromIter instead (in the Text related types), but that's a breaking change I've been avoiding.